src/aspire/denoising/denoiser_cov2d.py

```diff
@@ -18,72 +20,115 @@ class DenoiserCov2D(Denoiser):
+        self.star_file = star_file
+        self.starfile_path = os.path.dirname(star_file)
+        self.starfile_name = os.path.basename(star_file)
 
+        self.ctf_idx = None
+        self.ctf_fb = None
         # Assign the CTF information and index for each image
         if ctf_info and not (src.filters is None):
             logger.info(f'Represent CTF filters in FB basis')
-            uniq_ctf_filters = list(set(src.filters))
-            # Create the indices of CTF filters from all images
-            self.ctf_idx = np.array([uniq_ctf_filters.index(f) for f in src.filters])
-            # Evaluate CTFs in the FFB basis
-            self.ctf_fb = [f.fb_mat(basis) for f in uniq_ctf_filters]
         else:
             logger.info(f'CTF filters are not included in Cov2D denoising')
             # set all CTF filters to an identity filter
             self.ctf_idx = np.zeros(self.nimg, dtype=int)
             self.ctf_fb = [blk_diag_eye(blk_diag_partition(RadialCTFFilter().fb_mat(basis)))]
 
-    def denoise(self, covar_opt=None):
+        self.metadata = src._metadata.copy(deep=True)
+        self._write_star_file()
+        self.mrcs_out = self._get_mrcs_out()
+
+    def denoise(self, covar_opt=None, overwrite=False):
         """
         Denoiser 2D images using Cov2D method
 
-        :param basis: The basis method to expand 2D images
         :param covar_opt: The option list for building Cov2D matrix
+        :param overwrite: The option to owerwrite mrcs files for denoised images or not
         """
+        src = self.src
 
-        self.coeffs_noise = self.basis.evaluate_t(self.imgs_noise)
-
-        cov2d = RotCov2D(self.basis)
+        # Initialize the rotationally invariant covariance matrix of 2D images
+        # A fixed batch size is used to go through each image
+        cov2d = BatchedRotCov2D(self.src, self.basis, batch_size=512)
+        if not self.ctf_info:
+            cov2d.ctf_idx = self.ctf_idx
+            cov2d.ctf_fb = self.ctf_fb
 
         default_opt = {'shrinker': 'frobenius_norm', 'verbose': 0, 'max_iter': 250,
             'iter_callback': [], 'store_iterates': False, 'rel_tolerance': 1e-12,
             'precision': 'float64'}
         covar_opt = fill_struct(covar_opt, default_opt)
-        mean_coeffs_est = cov2d.get_mean(self.coeffs_noise, self.ctf_fb, self.ctf_idx)
+        # Calculate the mean and covariance for the rotationally invariant covariance matrix of 2D images
+        mean_coeffs_est = cov2d.get_mean()
+
+        covar_coeffs_est = cov2d.get_covar(noise_var=self.var_noise, mean_coeff=mean_coeffs_est,
+                                           covar_est_opt=covar_opt)
+        # Denoise the 2D images using the SPCAs from the rotationally invariant covariance matrix
+        # The batch size varies to fit the number of noisy images in each mrcs file
+        _, batch_size = src.group_unique_hist
+
+        img_start = 0
+        for ibatch in range(0, len(batch_size)):
```

Comment:
Why is this all done here? The images should be denoised *on demand* as part of a new `ImageSource`, not all at once.

Reply:
I tried to use the batch size according to the number of noisy images in each mrcs file. In the way, we can use multi-process mode easily in future by assigning one batch (one set of images in the same mrcs file) to one process and no need to worry about the order of file writing. I probably will change the function in the `ImageSource`  after the discussion.

Reply:
A `DenoisedImageSource` class was added for this purpose in the new commit 23ca25c.